### PR TITLE
feat(hyd): Simpler max step updater with equal time steps

### PR DESCRIPTION
**Pull request by crocket63**
## Summary of Changes

Before, a 320ms max time step updater would do 3 loops of 100ms and complete with one loop of 20ms to achieve a 320ms simulation update.

To improve as much as possible computation stability, this is now changed to have equal time steps each sub frame.
320ms second update will now result in 4 loops of 80ms. 

## Additional context

Discord username (if different from GitHub):

## Testing instructions

Not much impact expected but ensure you get correct behaviour of:
RAT rpm when deployed
Blue emergency generator RPM when in use (while RAT is out)
Cargo door opening/closing sequence still looks super awesome 

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: src/systems/systems/src/hydraulic/update_iterator.rs [hunk 1]
```diff
@@ -281,9 +275,14 @@ mod max_step_tests {
         for cur_time_step in &mut max_step {
             time_simulated += cur_time_step;
             actual_loop_num += 1;
+            println!(
+                "Loop done {} time step {}",
+                actual_loop_num,
+                cur_time_step.as_secs_f64()
+            );
```
Comment (beheh) [inline src/systems/systems/src/hydraulic/update_iterator.rs:282]: This log statement should probably not be commited?

Comment (crocket63) [reply src/systems/systems/src/hydraulic/update_iterator.rs:282]: That's in unit tests only, but yeah not mandatory anyway

File: src/systems/systems/src/hydraulic/update_iterator.rs [hunk 2]
```diff
@@ -283,7 +277,7 @@ mod max_step_tests {
             actual_loop_num += 1;
         }
 
-        //0.320 seconds with max of 0.100 we expect 3 max step duration plus 1 final step so 4
+        //0.320 seconds with max of 0.100 we expect 4 steps
```
Comment (davidwalschots) [inline src/systems/systems/src/hydraulic/update_iterator.rs:280]: **thought**: The fact that there is not a single test that changes despite the changes in this file means that more tests should be added for the `MaxFixedStepLoop`.
**suggestion**: Add various tests that confirm the durations returned by the iterator are in fact as you'd expect them to be.

File: src/systems/systems/src/hydraulic/update_iterator.rs [hunk 3]
```diff
@@ -71,50 +71,44 @@ impl Iterator for FixedStepLoop {
 /// Provides maximum fixed time interval looping.
 ///
 /// ## Example scenario
-/// With a max fixed time interval of 10 ms and a frame delta of 35 ms, this type will provide three
-/// iterations of 10 ms, and one iteration of 5ms to complete the 35ms total delta.
+/// With a max fixed time interval of 10 ms and a frame delta of 35 ms, this type will provide four
+/// iterations of 8.75ms, thus completing the 35ms total delta.
 #[derive(Copy, Clone)]
 pub struct MaxFixedStepLoop {
```
Comment (davidwalschots) [inline src/systems/systems/src/hydraulic/update_iterator.rs:77]: **thought**: I thought `Fixed` meant the time for most loops was equal. What did and does `Fixed` refer to if not that?

File: src/systems/systems/src/hydraulic/update_iterator.rs [hunk 4]
```diff
@@ -71,50 +71,44 @@ impl Iterator for FixedStepLoop {
...
-
-        self.num_of_max_step_loop = number_of_steps.floor() as u32;
-
-        let remaining_time_step_update = Duration::from_secs_f64(
-            (number_of_steps - (self.num_of_max_step_loop as f64)) * max_fixed_seconds,
-        );
-
-        if remaining_time_step_update > Duration::from_secs(0) {
-            self.remaining_frame_duration = Some(remaining_time_step_update);
+            self.frame_duration = Some(context.delta() / self.num_of_loops);
```
[the remark this answers is not in the document]
Comment (crocket63) [reply src/systems/systems/src/hydraulic/update_iterator.rs:96]: Yeah for sure but as long as it remains in the 10e-5/6/7/8, we should not be worried. Duration is based on f64 types underneath isn't it?

File: src/systems/systems/src/hydraulic/update_iterator.rs (hunk 3 above)
Comment (crocket63) [reply src/systems/systems/src/hydraulic/update_iterator.rs:77]: Yeah it's not Fixed it's MaxFixed: it's the Max that is fixed... Maybe MaxStepLoop is enough as a name then?

File: src/systems/systems/src/hydraulic/update_iterator.rs (hunk 2 above)
Comment (crocket63) [reply src/systems/systems/src/hydraulic/update_iterator.rs:280]: I actually check number of loops and final total time, current implementation and previous one gives different delta times but "satisfy" the intitial requirement.  The test is not checking the actual implementation though indeed.

File: src/systems/systems/src/hydraulic/update_iterator.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidwalschots) [reply src/systems/systems/src/hydraulic/update_iterator.rs:96]: Duration uses a `u64` and `u32` internally: https://doc.rust-lang.org/stable/src/core/time.rs.html#72-75

**Comment by davidwalschots** [comment]
After talking with @crocket63 we decided that I'd help him with getting the testing of this change sorted. I've added a `DeltaContext` trait which is implemented by both a `TestUpdateContext` as well as the regular `UpdateContext`. By doing this the tests can simply create a `TestUpdateContext` and do what they need to do. One doesn't need the whole `SimulationTestBed` for this.

**Review by davidwalschots — APPROVED** [review]

**Comment by mico975** [comment]
Quality Assurance Tester Report

Discord          : mico#3145
Object of testing: #6634
Tier of Testing  : 1
Date             : 05/02/2022

Testing Process: 
RAT, emerg elec gen rpm and blue hyd pres are all good and same as in previous RAT PR.
Cargo door behaviour still awesome.

Negatives:       
N/A

Testing Results: 
Passed

Conclusions: 
LGTM

**Pull request merged**